expansion.bzl

```diff
@@ -0,0 +1,561 @@
+"""
+Module containing functions that aid in environment variable expansion.
+
+This could be replaced with the same impl via Skylib dependency, if/when PR is merged.
+See https://github.com/bazelbuild/bazel-skylib/pull/486.
+"""
+
+_CONSIDERED_KEY_FORMATS = ("${}", "${{{}}}", "$({})")
+
+def _valid_char_for_env_var_name(char):
+    return char.isalnum() or char == "_"
+
+def _find_env_var_name_index_index(
+        string,
+        str_len,
+        search_start,
+        special_ending_char = None):
+    for offset in range(str_len - search_start):
+        index = search_start + offset
+        char = string[index]
+        if special_ending_char:
+            if char == special_ending_char:
+                return index
+        elif not _valid_char_for_env_var_name(char):
+            return index - 1
+    return str_len - 1
+
+def _odd_count_dollar_sign_repeat(containing_str, end_of_dollar_signs_index):
+    dollar_sign_count = 0
+    for index in range(end_of_dollar_signs_index, -1, -1):
+        if containing_str[index] != "$":
+            break
+        dollar_sign_count += 1
+    return (dollar_sign_count % 2) == 1
+
+def _key_to_be_expanded(str_with_key, key, start_of_key_index):
+    # Check that the string at index is prefixed with an odd number of `$`.
+    # Odd number means that the last `$` is not escaped.
+    odd_count = _odd_count_dollar_sign_repeat(str_with_key, start_of_key_index)
+
+    if not odd_count:
+        return False
+
+    # Check that the key is correctly matched.
+    # Specifically, check the key isn't matching to another key (substring).
+    key_mismatch = False
+    if key[-1] not in (")", "}"):
+        index_after_key = start_of_key_index + len(key)
+        key_mismatch = (
+            (index_after_key < len(str_with_key)) and
+            _valid_char_for_env_var_name(str_with_key[index_after_key])
+        )
+
+    return not key_mismatch
+
+def _fail_validation(fail_instead_of_return, found_errors_list, failure_message):
```

Comment:
fail_instead_of_return seems like an optional.

Reply:
I wanted to avoid optional arguments in the private functions. IMO, optional args are for flexibility for the caller, which is nice for the exposed API. But for private functionality, I'd prefer to have everything be explicit and clear (and there's no reason to have the flexibility, as the calls will all be similar and scoped to this one file).